docs/src/apalache/fold.md

```diff
@@ -0,0 +1,213 @@
+                                IN IF newCounters[e] > newCounters[currentMode]
+                                   THEN << newCounters, e >>
+                                   ELSE << newCounters, currentMode >>
+                           IN FoldSeq( CountElem, <<[ x \in ExtRange |-> 0 ], elIfEmpty >>, seq )[2]
+
+\* Returns TRUE iff fn is injective
+IsInjective(fn) == 
+  LET SeenBefore( seenAndResult, e ) == 
+    IF fn[e] \in seenAndResult[1]
+    THEN [ seenAndResult EXCEPT ![2] = FALSE ]
+    ELSE [ seenAndResult EXCEPT ![1] = seenAndResult[1] \union {fn[e]} ]
+  IN FoldSet( SeenBefore, << {}, TRUE >>, DOMAIN fn )[2]
+```
+
+For the sake of comparison, here are the same operators, using recursion, `CHOOSE` or quantification:
+
+```
+RECURSIVE Sum(_)
+Sum(S) == IF S = {} 
+          THEN 0
+          ELSE LET x == CHOOSE y \in S : TRUE
+               IN  x + Sum(S \ {x})
+
+RECURSIVE BigUnion(_)
+BigUnion(setOfSets) == IF setOfSets = {} 
+                       THEN {}
+                       ELSE LET S == CHOOSE x \in setOfSets : TRUE
+                            IN  S \union BigUnion(setOfSets \ {x})
+
+RECURSIVE SelectSeq(_,_)
+SelectSeq(seq, Test(_)) == IF seq = <<>>
+                           THEN <<>>
+                           ELSE LET tail == SelectSeq(Tail(seq), Test)
+                                IN IF Test( Head(seq) )
+                                   THEN <<Head(seq)>> \o tail
+                                   ELSE tail
+
+RECURSIVE Quantify(_,_)
+Quantify(S, P(_)) == IF S = {} 
+                     THEN 0
+                     ELSE LET x == CHOOSE y \in S : TRUE
+                          IN (IF P(x) THEN 1 ELSE 0) + Quantify(S \ {x}, P) 
+
+RECURSIVE Range(_)
+Range(seq) == IF seq = <<>>
+              THEN {}
+              ELSE {Head(seq)} \union Range(Tail(seq))
+
+Mode(seq, elIfEmpty) == IF seq = <<>>
+                        THEN elIfEmpty 
+                        ELSE LET numOf(p) == Quantify( DOMAIN seq, LAMBDA q: q = p )
+                             IN CHOOSE x \in Range(seq): \A y \in Range(seq) : numOf(x) >= numOf(y)
+
+IsInjective(fn) == \A a,b \in DOMAIN fn : fn[a] = fn[b] => a = b
+```
+
+In most cases, recursive operators are much more verbose, and the operators using `CHOOSE` and/or quantification mask double iteration (and thus have quadratic complexity). 
+For instance, the evaluation of the fold-less `IsInjective` operator actually requires the traversal of all domain pairs, instead of the single domain traversal with fold.
+In particular, `Mode`, the most verbose among the fold-defined operators, is still very readable (most LET-IN operators are introduced to improve readability, at the cost of verbosity) and quite efficient, as its complexity is linear w.r.t. the length of the sequence (the mode could also be computed directly, without a sub-call to `Range`, but the example would be more difficult to read), unlike the variant with `CHOOSE` and `\forall`, which is quadratic.
```

Comment:
```suggestion
In particular, `Mode`, the most verbose among the fold-defined operators, is still very readable (most LET-IN operators are introduced to improve readability, at the cost of verbosity) and quite efficient, as its complexity is linear w.r.t. the length of the sequence (the mode could also be computed directly, without a sub-call to `Range`, but the example would be more difficult to read), unlike the variant with `CHOOSE` and `\A`, which is quadratic.
```

Reply:
Why don't you give two versions with folds and explain which one would be better for Apalache? I believe that people are not writing TLA+ specs for efficiency. So you should label the more efficient version as efficient from the SMT encoding point of view, as there is no built-in notion of complexity in TLA+.

Reply:
Makes sense. They will care about efficiency, if they want to use automatic tools though.